src/full_node/blockchain.py

```diff
@@ -1,82 +1,77 @@
 from src.types.sized_bytes import bytes32
 from src.full_node.blockchain_check_conditions import blockchain_check_conditions_dict
+from src.full_node.sub_block_record import SubBlockRecord
+from src.types.unfinished_block import UnfinishedBlock
 from src.util.clvm import int_from_bytes
 from src.util.condition_tools import pkm_pairs_for_conditions_dict
 from src.full_node.cost_calculator import calculate_cost_of_program
-from src.util.errors import ConsensusError, Err
+from src.util.errors import Err
 from src.util.hash import std_hash
-from src.util.ints import uint32, uint64
-from src.util.merkle_set import MerkleSet
+from src.util.ints import uint32, uint64, uint8
+from src.full_node.block_root_validation import validate_block_merkle_roots
 from src.consensus.find_fork_point import find_fork_point_in_chain
+from src.consensus.block_rewards import calculate_pool_reward, calculate_base_farmer_reward
+from src.consensus.coinbase import create_pool_coin, create_farmer_coin
+from src.types.name_puzzle_condition import NPC
+from src.full_node.block_header_validation import validate_finished_header_block, validate_unfinished_header_block
+from src.types.unfinished_header_block import UnfinishedHeaderBlock
 
 log = logging.getLogger(__name__)
 
 
 class ReceiveBlockResult(Enum):
     """
     When Blockchain.receive_block(b) is called, one of these results is returned,
-    showing whether the block was added to the chain (extending a head or not),
+    showing whether the block was added to the chain (extending the peak),
     and if not, why it was not added.
     """
 
-    ADDED_TO_HEAD = 1  # Added to one of the heads, this block is now a new head
-    ADDED_AS_ORPHAN = 2  # Added as an orphan/stale block (block that is not a head or ancestor of a head)
+    NEW_PEAK = 1  # Added to the peak of the blockchain
+    ADDED_AS_ORPHAN = 2  # Added as an orphan/stale block (not a new peak of the chain)
     INVALID_BLOCK = 3  # Block was not added because it was invalid
     ALREADY_HAVE_BLOCK = 4  # Block is already present in this blockchain
-    DISCONNECTED_BLOCK = (
-        5  # Block's parent (previous pointer) is not in this blockchain
-    )
+    DISCONNECTED_BLOCK = 5  # Block's parent (previous pointer) is not in this blockchain
 
 
 class Blockchain:
     constants: ConsensusConstants
-    # Tips of the blockchain
-    tips: List[Header]
-    # Least common ancestor of tips
-    lca_block: Header
-    # Defines the path from genesis to the lca
+    # peak of the blockchain
+    peak_height: Optional[uint32]
+    # Defines the path from genesis to the peak, no orphan sub-blocks
     height_to_hash: Dict[uint32, bytes32]
-    # All headers (but not orphans) from genesis to the tip are guaranteed to be in headers
-    headers: Dict[bytes32, Header]
-    # Genesis block
-    genesis: FullBlock
+    # All sub blocks in peak path are guaranteed to be included, can include orphan sub-blocks
+    sub_blocks: Dict[bytes32, SubBlockRecord]
```

Comment:
@mariano54 
1. do we ever delete from this Dict ? i couldn't find where
2. does get_sub_blocks get all sub_blocks in history?
3. if the answer to 2 is yes, is this necessary ?

Reply:
it's never deleted. In the previous consensus it was fine, as the records were smaller, and there was 1 per 5 min in average.

with this consensus, it might require a large amount of memory right?

the difficulty with deleting old ones is that we still need to handle a long reorg up to genesis

Reply:
if we can handle long reorgs by re reading the subblocks from disk, instead of always having them in memory, then that's fine, we dont need all blocks here. We should ateast have all the hashes though, so we can find fork points.

Reply:
nice, that's something i can take care of, if we are building this thing to run for years we should take care of this